### PR TITLE
En 4551 trie node refactoring

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**
Add db, marshalizer and hasher to nodes instead of passing them around as parameters to functions.

### CODE REVIEW

File: data/trie/leafNode.go [hunk 1]
```diff
@@ -76,51 +90,75 @@ func (ln *leafNode) isDirty() bool {
...
 	err := ln.isEmptyOrNil()
 	if err != nil {
 		return nil, err
 	}
-	return encodeNodeAndGetHash(ln, marshalizer, hasher)
+	return encodeNodeAndGetHash(ln)
 }
 
-func (ln *leafNode) commit(force bool, level byte, originDb data.DBWriteCacher, targetDb data.DBWriteCacher, marshalizer marshal.Marshalizer, hasher hashing.Hasher) error {
+func (ln *leafNode) commit(force bool, level byte, db data.DBWriteCacher) error {
```
Comment (bogdan-rosianu) [inline data/trie/leafNode.go]: `targetDb` instead of `db`

File: data/trie/extensionNode.go [hunk 2]
```diff
@@ -110,62 +148,62 @@ func (en *extensionNode) setHash(marshalizer marshal.Marshalizer, hasher hashing
...
 			return nil, err
 		}
 		en.EncodedChild = encChild
 	}
-	return encodeNodeAndGetHash(en, marshalizer, hasher)
+	return encodeNodeAndGetHash(en)
 }
 
-func (en *extensionNode) commit(force bool, level byte, originDb data.DBWriteCacher, targetDb data.DBWriteCacher, marshalizer marshal.Marshalizer, hasher hashing.Hasher) error {
+func (en *extensionNode) commit(force bool, level byte, db data.DBWriteCacher) error {
```
Comment (bogdan-rosianu) [inline data/trie/extensionNode.go]: `targetDb` instead of `db`

File: data/trie/branchNode.go [hunk 3]
```diff
@@ -63,7 +63,35 @@ func branchNodeCapnToGo(src capnp.BranchNodeCapn, dest *branchNode) *branchNode
...
+	}
+	if marshalizer == nil || marshalizer.IsInterfaceNil() {
+		return nil, ErrNilMarshalizer
+	}
+	if hasher == nil || hasher.IsInterfaceNil() {
+		return nil, ErrNilHasher
+	}
+
+	var children [nrOfChildren]node
+	EncChildren := make([][]byte, nrOfChildren)
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply data/trie/branchNode.go]: Done.

File: data/trie/branchNode.go [hunk 4]
```diff
@@ -227,24 +280,24 @@ func (bn *branchNode) hashChildren(marshalizer marshal.Marshalizer, hasher hashi
...
 			}
 			bn.EncodedChildren[i] = encChild
 		}
 	}
-	return encodeNodeAndGetHash(bn, marshalizer, hasher)
+	return encodeNodeAndGetHash(bn)
 }
 
-func (bn *branchNode) commit(force bool, level byte, originDb data.DBWriteCacher, targetDb data.DBWriteCacher, marshalizer marshal.Marshalizer, hasher hashing.Hasher) error {
+func (bn *branchNode) commit(force bool, level byte, db data.DBWriteCacher) error {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply data/trie/branchNode.go]: Done.

File: data/trie/extensionNode.go [hunk 5]
```diff
@@ -287,12 +325,12 @@ func (en *extensionNode) getNext(key []byte, db data.DBWriteCacher, marshalizer
...
-func (en *extensionNode) insert(n *leafNode, db data.DBWriteCacher, marshalizer marshal.Marshalizer) (bool, node, [][]byte, error) {
+func (en *extensionNode) insert(n *leafNode) (bool, node, [][]byte, error) {
 	err := en.isEmptyOrNil()
 	if err != nil {
 		return false, nil, [][]byte{}, err
 	}
-	err = resolveIfCollapsed(en, 0, db, marshalizer)
+	err = resolveIfCollapsed(en, 0)
 	if err != nil {
 		return false, nil, [][]byte{}, err
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply data/trie/extensionNode.go]: Done.

File: data/trie/leafNode.go (hunk 1 above)
Comment (BeniaminDrasovean) [reply data/trie/leafNode.go]: Done.

File: data/trie/extensionNode.go (hunk 2 above)
Comment (BeniaminDrasovean) [reply data/trie/extensionNode.go]: Done.

**Pull request merged**